src/protocols/echo/Echo.h

```diff
@@ -0,0 +1,136 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP Echo unsolicitied
+ *      initaitor (client) and responder (server).
+ *
+ */
+
+#ifndef CHIP_ECHO_CURRENT_H_
+#define CHIP_ECHO_CURRENT_H_
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/CHIPProtocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+
+namespace chip {
+namespace Protocols {
+
+enum
+{
+    kEchoMessageType_EchoRequest  = 1,
+    kEchoMessageType_EchoResponse = 2
+};
+
+class DLL_EXPORT EchoClient : public ExchangeContextDelegate
+{
+public:
+    EchoClient(void);
+
+    /**
+     *  Initialize the EchoClient object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    exchangeMgr    A pointer to the ExchangeManager object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     */
+    CHIP_ERROR Init(ExchangeManager * exchangeMgr);
+
+    /**
+     *  Shutdown the EchoClient. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     *  @return #CHIP_NO_ERROR unconditionally.
+     */
+    CHIP_ERROR Shutdown(void);
```

Comment:
shoud this return void if it cannot fail?

Even if it could fail ... what would the recourse be? error here would be like throwing in a destructor ... sounds bad.

Reply:
Acked